### PR TITLE
BLS example with Rust integration

### PR DESCRIPTION
**Pull request by klkvr**
Adds a `BLSMultisig` contract demonstrating a mutlisignature account implementation controlled by BLS keys along with example demonstrating it integration in Rust with alloy and blst lib. 

It doesn't look like there's a clean way to sign arbitrary points through blst without `unsafe` blocks, so there is a decent amount of them

### CODE REVIEW

File: examples/rust-multisig/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "rust-multisig"
```
Comment (mattsse) [inline examples/rust-multisig/Cargo.toml:2]: maybe
```suggestion
name = "alphanet-multisig"
```

File: examples/rust-multisig/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    let mut agg_sig = signatures.swap_remove(0);
+    for sig in signatures {
+        unsafe { blst_p2_add_or_double(&mut agg_sig, &agg_sig, &sig) };
+    }
+
+    agg_sig.into()
+}
+
+#[tokio::main]
+pub async fn main() {
```
Comment (mattsse) [inline examples/rust-multisig/src/main.rs:152]: this is pretty cool, amazing that this works with anvil

**Review by mattsse — COMMENTED** [review]
this is great!

I'm not that familiar with the blst API, so would appreciate if @fgimenez could take a look

**Comment by klkvr** [comment]
Added example in Python as well, does pretty much the same thing

File: src/BLSMultisig.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {BLS} from "./sign/BLS.sol";
+
+/// @notice BLS-powered multisignature wallet, demonstrating the use of
+/// aggregated BLS signatures for verification
+/// @dev This is for demonstration purposes only, do not use in production. This contract does
+/// not include protection from rogue public-key attacks. You
```
Comment (fgimenez) [inline src/BLSMultisig.sol:9]:
```suggestion
/// not include protection from rogue public-key attacks. 
```

**Review by fgimenez — APPROVED** [review]
lgtm

File: src/sign/BLS.sol [hunk 4]
```diff
@@ -6,6 +6,8 @@ pragma solidity ^0.8.23;
 ///         defined in EIP-2537, see <https://eips.ethereum.org/EIPS/eip-2537>.
 /// @dev Precompile addresses come from the BLS addresses submodule in AlphaNet, see
 ///      <https://github.com/paradigmxyz/alphanet/blob/main/crates/precompile/src/addresses.rs>
+/// @notice `hashToCurve` logic is based on <https://github.com/ethyla/bls12-381-hash-to-curve/blob/main/src/HashToCurve.sol>
+/// with small modifications.
```
Comment (fgimenez) [inline src/sign/BLS.sol:10]: could you pls mention the modifications made?

**Review by fgimenez — COMMENTED** [review]
maybe not for this PR but it would be very nice to be able to run these examples on CI to make sure they keep working

**Comment by klkvr** [comment]
> maybe not for this PR but it would be very nice to be able to run these examples on CI to make sure they keep working, wdyt?

yep definitely, let's do it in a follow-up

**Review by mattsse — APPROVED** [review]

File: examples/bls-multisig/rust/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+use BLS::G2Point;
+
+sol! {
+    #[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+    #[sol(rpc)]
+    BLSMultisig,
+    "../../../out/BLSMultisig.sol/BLSMultisig.json"
+}
+
+impl From<[u8; 96]> for BLS::G1Point {
```
Comment (gakonst) [inline examples/bls-multisig/rust/src/main.rs:13]: makes me think we should create an alloy-bls package which also includes a tx signer for the AlphaNet payloads for the BatchContract -- cc @jxom

**Pull request merged**